File: src/dere_shared/emotion/decay.py
from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Literal

from loguru import logger

from dere_shared.emotion.models import (
    EMOTION_CHARACTERISTICS,
    EmotionCharacteristics,
    EmotionInstance,
    OCCEmotionType,
)


@dataclass
class EmotionDecayProfile:
    """Decay characteristics for an emotion type"""

    base_decay_rate: float  # Base decay rate per minute
    half_life: float  # Minutes for intensity to halve
    minimum_persistence: float  # Minimum time before any decay starts (minutes)
    resilience: float  # How much emotion resists decay (0-1)
    context_sensitivity: float  # How much external context affects decay (0-1)


# Decay profiles based on psychological research
EMOTION_DECAY_PROFILES: dict[OCCEmotionType, EmotionDecayProfile] = {
    # Positive emotions - generally decay faster but some are sticky
    OCCEmotionType.JOY: EmotionDecayProfile(0.08, 12, 2, 0.3, 0.6),
    OCCEmotionType.HOPE: EmotionDecayProfile(0.04, 25, 5, 0.6, 0.4),
    OCCEmotionType.SATISFACTION: EmotionDecayProfile(0.06, 18, 3, 0.4, 0.5),
    OCCEmotionType.RELIEF: EmotionDecayProfile(0.12, 8, 1, 0.2, 0.7),
    OCCEmotionType.HAPPY_FOR: EmotionDecayProfile(0.07, 15, 2, 0.3, 0.8),
    OCCEmotionType.PRIDE: EmotionDecayProfile(0.03, 30, 10, 0.7, 0.3),
    OCCEmotionType.ADMIRATION: EmotionDecayProfile(0.05, 20, 3, 0.4, 0.6),
    OCCEmotionType.LOVE: EmotionDecayProfile(0.01, 60, 15, 0.9, 0.2),
    OCCEmotionType.GRATITUDE: EmotionDecayProfile(0.04, 25, 5, 0.6, 0.5),
    OCCEmotionType.GRATIFICATION: EmotionDecayProfile(0.06, 18, 4, 0.4, 0.4),
    OCCEmotionType.INTEREST: EmotionDecayProfile(0.09, 10, 1, 0.3, 0.8),
    # Negative emotions - tend to be stickier
    OCCEmotionType.DISTRESS: EmotionDecayProfile(0.03, 30, 8, 0.7, 0.5),
    OCCEmotionType.FEAR: EmotionDecayProfile(0.02, 40, 10, 0.8, 0.3),
    OCCEmotionType.DISAPPOINTMENT: EmotionDecayProfile(0.05, 22, 5, 0.5, 0.6),
    OCCEmotionType.FEARS_CONFIRMED: EmotionDecayProfile(0.02, 45, 12, 0.8, 0.3),
    OCCEmotionType.PITY: EmotionDecayProfile(0.06, 18, 3, 0.4, 0.7),
    OCCEmotionType.GLOATING: EmotionDecayProfile(0.1, 7, 1, 0.2, 0.8),
    OCCEmotionType.RESENTMENT: EmotionDecayProfile(0.02, 50, 15, 0.8, 0.4),
    OCCEmotionType.SHAME: EmotionDecayProfile(0.02, 45, 12, 0.8, 0.3),
    OCCEmotionType.REPROACH: EmotionDecayProfile(0.04, 25, 6, 0.6, 0.5),
    OCCEmotionType.HATE: EmotionDecayProfile(0.01, 80, 20, 0.9, 0.2),
    OCCEmotionType.ANGER: EmotionDecayProfile(0.06, 18, 4, 0.5, 0.6),
    OCCEmotionType.REMORSE: EmotionDecayProfile(0.03, 35, 10, 0.7, 0.4),
    OCCEmotionType.DISGUST: EmotionDecayProfile(0.05, 20, 4, 0.5, 0.6),
    # Neutral
    OCCEmotionType.NEUTRAL: EmotionDecayProfile(0.15, 5, 0, 0.1, 0.9),
}


@dataclass
class DecayContext:
    """Context factors affecting decay rates"""

    is_user_present: bool
    is_user_engaged: bool
    recent_emotional_activity: float  # 0-1
    environmental_stress: float  # 0-1
    social_support: float  # 0-1
    time_of_day: Literal["morning", "afternoon", "evening", "night"]
    personality_stability: float  # 0-1


DEFAULT_DECAY_CONTEXT = DecayContext(
    is_user_present=False,
    is_user_engaged=False,
    recent_emotional_activity=0.5,
    environmental_stress=0.3,
    social_support=0.5,
    time_of_day="afternoon",
    personality_stability=0.6,
)


@dataclass
class DecayResult:
    """Result of decay calculation"""

    new_intensity: float
    decay_amount: float
    reasoning: str
    should_remove: bool


class SmartDecay:
    """Smart decay system with context-aware emotion persistence"""

    def __init__(
        self,
        emotion_decay_profiles: dict[OCCEmotionType, EmotionDecayProfile] = EMOTION_DECAY_PROFILES,
    ):
        self.emotion_decay_profiles = emotion_decay_profiles

    def calculate_decay(
        self,
        emotion: EmotionInstance,
        time_delta_minutes: float,
        context: DecayContext = DEFAULT_DECAY_CONTEXT,
    ) -> DecayResult:
        """Calculate decay for a single emotion"""
        if emotion.type == "neutral":
            return DecayResult(
                new_intensity=0,
                decay_amount=emotion.intensity,
                reasoning="Neutral emotion removed",
                should_remove=True,
            )

        profile = self.emotion_decay_profiles[emotion.type]  # type: ignore
        characteristics = EMOTION_CHARACTERISTICS[emotion.type]  # type: ignore

        emotion_age = time_delta_minutes

        # Check minimum persistence
        if emotion_age < profile.minimum_persistence:
            return DecayResult(
                new_intensity=emotion.intensity,
                decay_amount=0,
                reasoning=f"Too recent ({emotion_age:.1f}m < {profile.minimum_persistence}m minimum)",
                should_remove=False,
            )

        # Calculate adjusted decay rate
        adjusted_decay_rate = self._calculate_adjusted_decay_rate(profile, characteristics, context)

        # Exponential decay
        base_decay_factor = math.exp(-adjusted_decay_rate * time_delta_minutes)
        new_intensity = emotion.intensity * base_decay_factor

        # Apply resilience
        resilience_protection = (emotion.intensity / 100) ** 0.5 * profile.resilience
        new_intensity = emotion.intensity - (emotion.intensity - new_intensity) * (
            1 - resilience_protection
        )

        # Apply contextual modifiers
        new_intensity = self._apply_contextual_modifiers(
            new_intensity, emotion.intensity, characteristics, context
        )

        new_intensity = max(0, min(100, new_intensity))
        decay_amount = emotion.intensity - new_intensity

        # Calculate removal threshold
        removal_threshold = self._calculate_removal_threshold(characteristics, context)
        should_remove = new_intensity < removal_threshold

        reasoning = self._generate_decay_reasoning(
            emotion.type,  # type: ignore
            emotion.intensity,
            new_intensity,
            adjusted_decay_rate,
            resilience_protection,
            context,
            emotion_age,
        )

        logger.debug(
            f"[SmartDecay] {emotion.type} decay: "
            f"{emotion.intensity:.1f} → {new_intensity:.1f}, "
            f"age={emotion_age:.1f}m, remove={should_remove}"
        )

        return DecayResult(
            new_intensity=0 if should_remove else new_intensity,
            decay_amount=decay_amount,
            reasoning=reasoning,
            should_remove=should_remove,
        )

    def apply_decay_to_emotions(
        self,
        emotions: dict[OCCEmotionType | str, EmotionInstance],
        time_delta_minutes: float,
        context: DecayContext = DEFAULT_DECAY_CONTEXT,
    ) -> dict:
        """Apply decay to multiple emotions"""
        updated_emotions: dict[OCCEmotionType | str, EmotionInstance] = {}
        decay_results: list[dict] = []
        total_decay_activity = 0.0

        for emotion_type, emotion in emotions.items():
            decay_result = self.calculate_decay(emotion, time_delta_minutes, context)
            decay_results.append({"type": emotion_type, "result": decay_result})

            total_decay_activity += decay_result.decay_amount

            if not decay_result.should_remove and decay_result.new_intensity > 0:
                import time

                updated_emotions[emotion_type] = EmotionInstance(
                    type=emotion.type,
                    intensity=decay_result.new_intensity,
                    last_updated=int(time.time() * 1000),
                )

        return {
            "updated_emotions": updated_emotions,
            "decay_results": decay_results,
            "total_decay_activity": total_decay_activity,
        }

    def _calculate_adjusted_decay_rate(
        self,
        profile: EmotionDecayProfile,
        characteristics: EmotionCharacteristics,
        context: DecayContext,
    ) -> float:
        """Calculate context-adjusted decay rate"""
        adjusted_rate = profile.base_decay_rate

        # User presence affects decay
        if not context.is_user_present:
            if characteristics.social_relevance == "high":
                adjusted_rate *= 1.3
            elif characteristics.social_relevance == "medium":
                adjusted_rate *= 1.1

        if context.is_user_engaged:
            adjusted_rate *= 0.8

        # Emotional activity affects decay
        if context.recent_emotional_activity > 0.7:
            adjusted_rate *= 0.7
        elif context.recent_emotional_activity < 0.3:
            adjusted_rate *= 1.2

        # Environmental stress
        if context.environmental_stress > 0.6:
            if characteristics.valence == "positive":
                adjusted_rate *= 1.4
            else:
                adjusted_rate *= 0.8

        # Social support
        if context.social_support > 0.6:
            if characteristics.valence == "positive":
                adjusted_rate *= 0.9
            elif characteristics.valence == "negative":
                adjusted_rate *= 1.2

        # Time of day
        match context.time_of_day:
            case "morning":
                adjusted_rate *= 1.1
            case "evening":
                adjusted_rate *= 0.9
            case "night":
                if characteristics.valence == "negative":
                    adjusted_rate *= 0.7

        # Personality stability
        stability_factor = 0.5 + context.personality_stability * 0.5
        adjusted_rate *= stability_factor

        return max(0.001, adjusted_rate)

    def _apply_contextual_modifiers(
        self,
        new_intensity: float,
        original_intensity: float,
        characteristics: EmotionCharacteristics,
        context: DecayContext,
    ) -> float:
        """Apply contextual modifications to decay"""
        modified_intensity = new_intensity

        # High arousal rebound
        if characteristics.arousal == "high" and context.recent_emotional_activity > 0.8:
            rebound_factor = 1.05
            modified_intensity = new_intensity + (original_intensity - new_intensity) * (
                rebound_factor - 1
            )

        # Sticky emotions resist decay more
        if characteristics.persistence == "sticky":
            if characteristics.valence == "positive" and context.social_support > 0.7:
                modified_intensity = new_intensity + (original_intensity - new_intensity) * 0.1
            elif characteristics.valence == "negative" and context.environmental_stress > 0.6:
                modified_intensity = new_intensity + (original_intensity - new_intensity) * 0.15

        return max(0, min(100, modified_intensity))

    def _calculate_removal_threshold(
        self, characteristics: EmotionCharacteristics, context: DecayContext
    ) -> float:
        """Calculate threshold for emotion removal"""
        threshold = 1.0

        if characteristics.persistence == "sticky":
            threshold = 0.5
        elif characteristics.persistence == "fleeting":
            threshold = 2.0

        threshold *= 0.5 + context.personality_stability * 0.5

        return threshold

    def _generate_decay_reasoning(
        self,
        emotion_type: OCCEmotionType,
        original_intensity: float,
        new_intensity: float,
        decay_rate: float,
        resilience_protection: float,
        context: DecayContext,
        age_minutes: float,
    ) -> str:
        """Generate human-readable decay reasoning"""
        parts: list[str] = []

        parts.append(f"{emotion_type}: {original_intensity:.1f} → {new_intensity:.1f}")
        parts.append(f"age: {age_minutes:.1f}m")
        parts.append(f"rate: {decay_rate:.3f}")

        if resilience_protection > 0.1:
            parts.append(f"resilience: {resilience_protection * 100:.0f}%")

        if not context.is_user_present:
            parts.append("user away")

        if context.recent_emotional_activity > 0.7:
            parts.append("high activity")
        elif context.recent_emotional_activity < 0.3:
            parts.append("low activity")

        return ", ".join(parts)

File: packages/daemon/src/emotions/decay.ts
import {
  EMOTION_CHARACTERISTICS,
  type EmotionCharacteristics,
  type EmotionInstance,
} from "./physics.js";

type EmotionDecayProfile = {
  base_decay_rate: number;
  half_life: number;
  minimum_persistence: number;
  resilience: number;
  context_sensitivity: number;
};

export type DecayContext = {
  is_user_present: boolean;
  is_user_engaged: boolean;
  recent_emotional_activity: number;
  environmental_stress: number;
  social_support: number;
  time_of_day: "morning" | "afternoon" | "evening" | "night";
  personality_stability: number;
};

export const DEFAULT_DECAY_CONTEXT: DecayContext = {
  is_user_present: false,
  is_user_engaged: false,
  recent_emotional_activity: 0.5,
  environmental_stress: 0.3,
  social_support: 0.5,
  time_of_day: "afternoon",
  personality_stability: 0.6,
};

type DecayResult = {
  new_intensity: number;
  decay_amount: number;
  reasoning: string;
  should_remove: boolean;
};

const EMOTION_DECAY_PROFILES: Record<string, EmotionDecayProfile> & {
  neutral: EmotionDecayProfile;
} = {
  joy: {
    base_decay_rate: 0.08,
    half_life: 12,
    minimum_persistence: 2,
    resilience: 0.3,
    context_sensitivity: 0.6,
  },
  hope: {
    base_decay_rate: 0.04,
    half_life: 25,
    minimum_persistence: 5,
    resilience: 0.6,
    context_sensitivity: 0.4,
  },
  satisfaction: {
    base_decay_rate: 0.06,
    half_life: 18,
    minimum_persistence: 3,
    resilience: 0.4,
    context_sensitivity: 0.5,
  },
  relief: {
    base_decay_rate: 0.12,
    half_life: 8,
    minimum_persistence: 1,
    resilience: 0.2,
    context_sensitivity: 0.7,
  },
  "happy-for": {
    base_decay_rate: 0.07,
    half_life: 15,
    minimum_persistence: 2,
    resilience: 0.3,
    context_sensitivity: 0.8,
  },
  pride: {
    base_decay_rate: 0.03,
    half_life: 30,
    minimum_persistence: 10,
    resilience: 0.7,
    context_sensitivity: 0.3,
  },
  admiration: {
    base_decay_rate: 0.05,
    half_life: 20,
    minimum_persistence: 3,
    resilience: 0.4,
    context_sensitivity: 0.6,
  },
  love: {
    base_decay_rate: 0.01,
    half_life: 60,
    minimum_persistence: 15,
    resilience: 0.9,
    context_sensitivity: 0.2,
  },
  gratitude: {
    base_decay_rate: 0.04,
    half_life: 25,
    minimum_persistence: 5,
    resilience: 0.6,
    context_sensitivity: 0.5,
  },
  gratification: {
    base_decay_rate: 0.06,
    half_life: 18,
    minimum_persistence: 4,
    resilience: 0.4,
    context_sensitivity: 0.4,
  },
  interest: {
    base_decay_rate: 0.09,
    half_life: 10,
    minimum_persistence: 1,
    resilience: 0.3,
    context_sensitivity: 0.8,
  },
  distress: {
    base_decay_rate: 0.03,
    half_life: 30,
    minimum_persistence: 8,
    resilience: 0.7,
    context_sensitivity: 0.5,
  },
  fear: {
    base_decay_rate: 0.02,
    half_life: 40,
    minimum_persistence: 10,
    resilience: 0.8,
    context_sensitivity: 0.3,
  },
  disappointment: {
    base_decay_rate: 0.05,
    half_life: 22,
    minimum_persistence: 5,
    resilience: 0.5,
    context_sensitivity: 0.6,
  },
  "fears-confirmed": {
    base_decay_rate: 0.02,
    half_life: 45,
    minimum_persistence: 12,
    resilience: 0.8,
    context_sensitivity: 0.3,
  },
  pity: {
    base_decay_rate: 0.06,
    half_life: 18,
    minimum_persistence: 3,
    resilience: 0.4,
    context_sensitivity: 0.7,
  },
  gloating: {
    base_decay_rate: 0.1,
    half_life: 7,
    minimum_persistence: 1,
    resilience: 0.2,
    context_sensitivity: 0.8,
  },
  resentment: {
    base_decay_rate: 0.02,
    half_life: 50,
    minimum_persistence: 15,
    resilience: 0.8,
    context_sensitivity: 0.4,
  },
  shame: {
    base_decay_rate: 0.02,
    half_life: 45,
    minimum_persistence: 12,
    resilience: 0.8,
    context_sensitivity: 0.3,
  },
  reproach: {
    base_decay_rate: 0.04,
    half_life: 25,
    minimum_persistence: 6,
    resilience: 0.6,
    context_sensitivity: 0.5,
  },
  hate: {
    base_decay_rate: 0.01,
    half_life: 80,
    minimum_persistence: 20,
    resilience: 0.9,
    context_sensitivity: 0.2,
  },
  anger: {
    base_decay_rate: 0.06,
    half_life: 18,
    minimum_persistence: 4,
    resilience: 0.5,
    context_sensitivity: 0.6,
  },
  remorse: {
    base_decay_rate: 0.03,
    half_life: 35,
    minimum_persistence: 10,
    resilience: 0.7,
    context_sensitivity: 0.4,
  },
  disgust: {
    base_decay_rate: 0.05,
    half_life: 20,
    minimum_persistence: 4,
    resilience: 0.5,
    context_sensitivity: 0.6,
  },
  neutral: {
    base_decay_rate: 0.15,
    half_life: 5,
    minimum_persistence: 0,
    resilience: 0.1,
    context_sensitivity: 0.9,
  },
};

export class SmartDecay {
  private readonly profiles: Record<string, EmotionDecayProfile> & {
    neutral: EmotionDecayProfile;
  };

  constructor(
    profiles: Record<string, EmotionDecayProfile> & {
      neutral: EmotionDecayProfile;
    } = EMOTION_DECAY_PROFILES,
  ) {
    this.profiles = profiles;
  }

  calculateDecay(
    emotion: EmotionInstance,
    timeDeltaMinutes: number,
    context: DecayContext = DEFAULT_DECAY_CONTEXT,
  ): DecayResult {
    if (emotion.type === "neutral") {
      return {
        new_intensity: 0,
        decay_amount: emotion.intensity,
        reasoning: "Neutral emotion removed",
        should_remove: true,
      };
    }

    const profile = this.profiles[emotion.type] ?? this.profiles.neutral;
    const characteristics =
      EMOTION_CHARACTERISTICS[emotion.type] ?? EMOTION_CHARACTERISTICS.neutral;

    const emotionAge = timeDeltaMinutes;
    if (emotionAge < profile.minimum_persistence) {
      return {
        new_intensity: emotion.intensity,
        decay_amount: 0,
        reasoning: `Too recent (${emotionAge.toFixed(1)}m < ${profile.minimum_persistence}m minimum)`,
        should_remove: false,
      };
    }

    const adjustedDecayRate = this.calculateAdjustedDecayRate(profile, characteristics, context);
    const baseDecayFactor = Math.exp(-adjustedDecayRate * timeDeltaMinutes);
    let newIntensity = emotion.intensity * baseDecayFactor;

    const resilienceProtection = Math.pow(emotion.intensity / 100, 0.5) * profile.resilience;
    newIntensity =
      emotion.intensity - (emotion.intensity - newIntensity) * (1 - resilienceProtection);

    newIntensity = this.applyContextualModifiers(
      newIntensity,
      emotion.intensity,
      characteristics,
      context,
    );

    newIntensity = Math.max(0, Math.min(100, newIntensity));
    const decayAmount = emotion.intensity - newIntensity;
    const removalThreshold = this.calculateRemovalThreshold(characteristics, context);
    const shouldRemove = newIntensity < removalThreshold;

    const reasoning = this.generateDecayReasoning(
      emotion.type,
      emotion.intensity,
      newIntensity,
      adjustedDecayRate,
      resilienceProtection,
      context,
      emotionAge,
    );

    return {
      new_intensity: shouldRemove ? 0 : newIntensity,
      decay_amount: decayAmount,
      reasoning,
      should_remove: shouldRemove,
    };
  }

  applyDecayToEmotions(
    emotions: Record<string, EmotionInstance>,
    timeDeltaMinutes: number,
    context: DecayContext = DEFAULT_DECAY_CONTEXT,
  ): {
    updated_emotions: Record<string, EmotionInstance>;
    decay_results: Array<{ type: string; result: DecayResult }>;
    total_decay_activity: number;
  } {
    const updated: Record<string, EmotionInstance> = {};
    const decayResults: Array<{ type: string; result: DecayResult }> = [];
    let totalDecayActivity = 0;
    const now = Date.now();

    for (const [emotionType, emotion] of Object.entries(emotions)) {
      const result = this.calculateDecay(emotion, timeDeltaMinutes, context);
      decayResults.push({ type: emotionType, result });
      totalDecayActivity += result.decay_amount;
      if (!result.should_remove && result.new_intensity > 0) {
        updated[emotionType] = {
          type: emotion.type,
          intensity: result.new_intensity,
          last_updated: now,
        };
      }
    }

    return {
      updated_emotions: updated,
      decay_results: decayResults,
      total_decay_activity: totalDecayActivity,
    };
  }

  private calculateAdjustedDecayRate(
    profile: EmotionDecayProfile,
    characteristics: EmotionCharacteristics,
    context: DecayContext,
  ): number {
    let adjusted = profile.base_decay_rate;

    if (!context.is_user_present) {
      if (characteristics.social_relevance === "high") {
        adjusted *= 1.3;
      } else if (characteristics.social_relevance === "medium") {
        adjusted *= 1.1;
      }
    }

    if (context.is_user_engaged) {
      adjusted *= 0.8;
    }

    if (context.recent_emotional_activity > 0.7) {
      adjusted *= 0.7;
    } else if (context.recent_emotional_activity < 0.3) {
      adjusted *= 1.2;
    }

    if (context.environmental_stress > 0.6) {
      if (characteristics.valence === "positive") {
        adjusted *= 1.4;
      } else {
        adjusted *= 0.8;
      }
    }

    if (context.social_support > 0.6) {
      if (characteristics.valence === "positive") {
        adjusted *= 0.9;
      } else if (characteristics.valence === "negative") {
        adjusted *= 1.2;
      }
    }

    if (context.time_of_day === "morning") {
      adjusted *= 1.1;
    } else if (context.time_of_day === "evening") {
      adjusted *= 0.9;
    } else if (context.time_of_day === "night" && characteristics.valence === "negative") {
      adjusted *= 0.7;
    }

    const stabilityFactor = 0.5 + context.personality_stability * 0.5;
    adjusted *= stabilityFactor;

    return Math.max(0.001, adjusted);
  }

  private applyContextualModifiers(
    newIntensity: number,
    originalIntensity: number,
    characteristics: EmotionCharacteristics,
    context: DecayContext,
  ): number {
    let modified = newIntensity;

    if (characteristics.arousal === "high" && context.recent_emotional_activity > 0.8) {
      const reboundFactor = 1.05;
      modified = newIntensity + (originalIntensity - newIntensity) * (reboundFactor - 1);
    }

    if (characteristics.persistence === "sticky") {
      if (characteristics.valence === "positive" && context.social_support > 0.7) {
        modified = newIntensity + (originalIntensity - newIntensity) * 0.1;
      } else if (characteristics.valence === "negative" && context.environmental_stress > 0.6) {
        modified = newIntensity + (originalIntensity - newIntensity) * 0.15;
      }
    }

    return Math.max(0, Math.min(100, modified));
  }

  private calculateRemovalThreshold(
    characteristics: EmotionCharacteristics,
    context: DecayContext,
  ): number {
    let threshold = 1.0;
    if (characteristics.persistence === "sticky") {
      threshold = 0.5;
    } else if (characteristics.persistence === "fleeting") {
      threshold = 2.0;
    }

    threshold *= 0.5 + context.personality_stability * 0.5;
    return threshold;
  }

  private generateDecayReasoning(
    emotionType: string,
    originalIntensity: number,
    newIntensity: number,
    decayRate: number,
    resilienceProtection: number,
    context: DecayContext,
    ageMinutes: number,
  ): string {
    const parts: string[] = [];
    parts.push(`${emotionType}: ${originalIntensity.toFixed(1)} -> ${newIntensity.toFixed(1)}`);
    parts.push(`age: ${ageMinutes.toFixed(1)}m`);
    parts.push(`rate: ${decayRate.toFixed(3)}`);

    if (resilienceProtection > 0.1) {
      parts.push(`resilience: ${(resilienceProtection * 100).toFixed(0)}%`);
    }

    if (!context.is_user_present) {
      parts.push("user away");
    }

    if (context.recent_emotional_activity > 0.7) {
      parts.push("high activity");
    } else if (context.recent_emotional_activity < 0.3) {
      parts.push("low activity");
    }

    return parts.join(", ");
  }
}
